File: E2E-AUTO-MICRO-APPS/integrate-memory-bank.js
/**
 * Memory-Bank Integration Protocol
 * 
 * Establishes bidirectional interfaces between Memory-Bank and other
 * system components to enable cognitive persistence across modules.
 */

const fs = require('fs');
const path = require('path');

// Module identity constants
const MODULE_NAME = 'memory-bank';
const MODULE_ROOT = path.join(__dirname, 'modules', MODULE_NAME);
const INTEGRATIONS_DIR = path.join(MODULE_ROOT, 'integrations');

// Component integration mappings
const INTEGRATION_COMPONENTS = [
  {
    name: 'openrouter',
    description: 'AI model provider integration',
    interfaces: ['contextual-retrieval', 'embedding-cache'],
    priority: 1
  },
  {
    name: 'task-master',
    description: 'Task management integration',
    interfaces: ['task-history', 'dependency-tracking'],
    priority: 2
  },
  {
    name: 'context7',
    description: 'Documentation context integration',
    interfaces: ['doc-cache', 'usage-metrics'],
    priority: 3
  },
  {
    name: 'n8n',
    description: 'Workflow automation integration',
    interfaces: ['workflow-state', 'execution-history'],
    priority: 4
  }
];

/**
 * Integration orchestration controller
 */
async function integrateMemoryBank() {
  console.log('╭─────────────────────────────────────────╮');
  console.log('│ MEMORY-BANK INTEGRATION PROTOCOL        │');
  console.log('│ Cognitive Persistence Interface Mapping │');
  console.log('╰─────────────────────────────────────────╯');

  try {
    // Phase 1: Create integration directory structure
    await createIntegrationStructure();
    
    // Phase 2: Generate integration interfaces
    await generateInterfaces();
    
    // Phase 3: Create adapter modules
    await createAdapters();
    
    // Phase 4: Register with component modules
    await registerWithComponents();
    
    console.log('\n✅ Memory-Bank integration sequence completed successfully');
    console.log('   Cognitive interfaces established with system components');
  } catch (error) {
    console.error('\n❌ Integration sequence failed:', error.message);
    console.error('   See detailed error information above');
    process.exit(1);
  }
}

/**
 * Creates the directory structure for integrations
 */
async function createIntegrationStructure() {
  console.log('\n📂 Creating integration structure...');
  
  // Create integrations directory
  if (!fs.existsSync(INTEGRATIONS_DIR)) {
    fs.mkdirSync(INTEGRATIONS_DIR, { recursive: true });
    console.log('  Created integrations directory ✓');
  }
  
  // Create component directories
  for (const component of INTEGRATION_COMPONENTS) {
    const componentDir = path.join(INTEGRATIONS_DIR, component.name);
    if (!fs.existsSync(componentDir)) {
      fs.mkdirSync(componentDir, { recursive: true });
      console.log(`  Created ${component.name} integration directory ✓`);
    }
  }
}

/**
 * Generates interface modules for each component
 */
async function generateInterfaces() {
  console.log('\n🔌 Generating integration interfaces...');
  
  for (const component of INTEGRATION_COMPONENTS) {
    console.log(`\n  Configuring ${component.name} interfaces:`);
    
    for (const interfaceType of component.interfaces) {
      const interfaceFile = path.join(INTEGRATIONS_DIR, component.name, `${interfaceType}.js`);
      
      if (!fs.existsSync(interfaceFile)) {
        // Generate interface file based on type
        const interfaceContent = generateInterfaceTemplate(component.name, interfaceType);
        fs.writeFileSync(interfaceFile, interfaceContent);
        console.log(`    Created ${interfaceType} interface ✓`);
      } else {
        console.log(`    ${interfaceType} interface already exists ✓`);
      }
    }
    
    // Create index.js to export all interfaces
    const indexFile = path.join(INTEGRATIONS_DIR, component.name, 'index.js');
    const indexContent = generateIndexFile(component);
    fs.writeFileSync(indexFile, indexContent);
    console.log(`    Created index file ✓`);
  }
}

/**
 * Creates adapter modules for connecting to components
 */
async function createAdapters() {
  console.log('\n🧩 Creating component adapters...');
  
  // Create main adapter index file
  const adaptersIndexPath = path.join(INTEGRATIONS_DIR, 'index.js');
  const adaptersIndexContent = generateAdaptersIndex(INTEGRATION_COMPONENTS);
  fs.writeFileSync(adaptersIndexPath, adaptersIndexContent);
  console.log('  Created main integration index file ✓');
}

/**
 * Registers memory-bank with target components
 */
async function registerWithComponents() {
  console.log('\n🔗 Registering with components...');
  
  // For each component, check if it exists and create registration
  for (const component of INTEGRATION_COMPONENTS) {
    const componentDir = path.join(__dirname, 'modules', component.name);
    
    if (fs.existsSync(componentDir)) {
      console.log(`  Registering with ${component.name}...`);
      
      // Create registration hook in component directory
      const registrationFile = path.join(componentDir, 'memory-bank-hook.js');
      const registrationContent = generateRegistrationHook(component);
      fs.writeFileSync(registrationFile, registrationContent);
      
      console.log(`    Created registration hook ✓`);
    } else {
      console.log(`  Component ${component.name} not found, skipping registration`);
    }
  }
}

/**
 * Generates template code for an interface file
 */
function generateInterfaceTemplate(component, interfaceType) {
  return `/**
 * Memory-Bank: ${component} ${interfaceType} Interface
 * 
 * Provides bidirectional integration between Memory-Bank
 * and the ${component} module for ${interfaceType} functionality.
 */

const MemoryBank = require('../../../index');

/**
 * Interface factory function
 * @param {Object} options - Configuration options
 * @returns {Object} Interface methods
 */
module.exports = function create${toCamelCase(interfaceType)}Interface(options = {}) {
  // Initialize memory bank instance or use provided one
  const memoryBank = options.memoryBank || new MemoryBank();
  
  // Define the category for this interface
  const category = '${component}-${interfaceType}';
  
  return {
    /**
     * Store data in memory bank
     * @param {string} key - Unique identifier
     * @param {any} data - Data to store
     * @param {Object} metadata - Additional metadata
     * @returns {Promise<boolean>} Success status
     */
    async store(key, data, metadata = {}) {
      try {
        return await memoryBank.store(category, key, data, { metadata });
      } catch (error) {
        console.error(\`[${component}-${interfaceType}] Store error:\`, error);
        return false;
      }
    },
    
    /**
     * Retrieve data from memory bank
     * @param {string} key - Unique identifier
     * @returns {Promise<any>} Retrieved data or null
     */
    async retrieve(key) {
      try {
        return await memoryBank.retrieve(category, key);
      } catch (error) {
        console.error(\`[${component}-${interfaceType}] Retrieve error:\`, error);
        return null;
      }
    },
    
    /**
     * List all keys in this category
     * @returns {Promise<Array<string>>} List of keys
     */
    async list() {
      try {
        return await memoryBank.list(category);
      } catch (error) {
        console.error(\`[${component}-${interfaceType}] List error:\`, error);
        return [];
      }
    },
    
    /**
     * Clear data by key
     * @param {string} key - Unique identifier
     * @returns {Promise<boolean>} Success status
     */
    async clear(key) {
      try {
        return await memoryBank.delete(category, key);
      } catch (error) {
        console.error(\`[${component}-${interfaceType}] Clear error:\`, error);
        return false;
      }
    }
  };
};
`;
}

/**
 * Generates an index file for a component's interfaces
 */
function generateIndexFile(component) {
  const imports = component.interfaces.map(interfaceType => 
    `const create${toCamelCase(interfaceType)}Interface = require('./${interfaceType}');`
  ).join('\n');
  
  const exports = `module.exports = {
${component.interfaces.map(interfaceType => 
    `  create${toCamelCase(interfaceType)}Interface,`
  ).join('\n')}
};`;

  return `/**
 * Memory-Bank: ${component.name} Integration
 * 
 * ${component.description}
 */

${imports}

${exports}
`;
}

/**
 * Generates the main adapters index file
 */
function generateAdaptersIndex(components) {
  const imports = components.map(component => 
    `const ${component.name}Interfaces = require('./${component.name}');`
  ).join('\n');
  
  const exports = `module.exports = {
${components.map(component => `  ${component.name}: ${component.name}Interfaces,`).join('\n')}
  
  /**
   * Initialize all integrations with a single memory bank instance
   * @param {Object} memoryBank - Memory bank instance
   * @returns {Object} Initialized interfaces grouped by component
   */
  initializeAll(memoryBank) {
    const integrations = {};
    
${components.map(component => `    integrations.${component.name} = {};
${component.interfaces.map(interfaceType => 
      `    integrations.${component.name}.${camelCaseToDash(interfaceType)} = ${component.name}Interfaces.create${toCamelCase(interfaceType)}Interface({ memoryBank });`
    ).join('\n')}`).join('\n\n')}
    
    return integrations;
  }
};`;

  return `/**
 * Memory-Bank: Integration Hub
 * 
 * Central access point for all component integrations
 */

${imports}

${exports}
`;
}

/**
 * Generates a registration hook for a component
 */
function generateRegistrationHook(component) {
  return `/**
 * Memory-Bank Registration Hook for ${component.name}
 * 
 * Enables ${component.name} to interface with Memory-Bank cognitive persistence
 */

const path = require('path');
const MemoryBank = require('../memory-bank');
const { ${component.name}: interfaces } = require('../memory-bank/integrations');

// Create memory-bank instance if not already available
const memoryBank = global.memoryBank || new MemoryBank();

// Register memory-bank globally if not already registered
if (!global.memoryBank) {
  global.memoryBank = memoryBank;
}

// Initialize interfaces
const ${component.name}Memory = {};

${component.interfaces.map(interfaceType => 
  `${component.name}Memory.${camelCaseToDash(interfaceType)} = interfaces.create${toCamelCase(interfaceType)}Interface({ memoryBank });`
).join('\n')}

module.exports = ${component.name}Memory;
`;
}

/**
 * Utility function to convert dash-case to camelCase
 */
function toCamelCase(str) {
  return str.split('-')
    .map((word, index) => 
      index === 0 ? word : word.charAt(0).toUpperCase() + word.slice(1)
    )
    .join('');
}

/**
 * Utility function to convert camelCase to dash-case
 */
function camelCaseToDash(str) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

// Execute integration sequence
integrateMemoryBank();

File: E2E-AUTO-MICRO-APPS/integrate_memory_bank.py
#!/usr/bin/env python3
"""
Memory-Bank Integration Protocol

This script initializes and integrates the Memory-Bank subsystem
into the E2E-AUTO-MICRO-APPS orchestration framework using a
modular integration approach with robust error handling.
"""

import os
import sys
import json
import shutil
from pathlib import Path

# Configuration
PROJECT_ROOT = Path('/home/tabs/E2E-AUTO-MICRO-APPS')
MODULES_DIR = PROJECT_ROOT / 'modules'
MEMORY_BANK_DIR = MODULES_DIR / 'memory-bank'
MEMORY_STORAGE_DIR = PROJECT_ROOT / '.memory'

def print_status(message, success=True):
    """Print a formatted status message"""
    prefix = "✅" if success else "❌"
    print(f"{prefix} {message}")

def ensure_directory(path):
    """Ensure a directory exists, creating it if necessary"""
    try:
        os.makedirs(path, exist_ok=True)
        return True
    except Exception as e:
        print_status(f"Failed to create directory {path}: {e}", False)
        return False

def write_file(path, content):
    """Write content to a file"""
    try:
        with open(path, 'w') as f:
            f.write(content)
        return True
    except Exception as e:
        print_status(f"Failed to write to {path}: {e}", False)
        return False

def verify_memory_bank():
    """Verify that the Memory-Bank module exists and has all required files"""
    print("Verifying Memory-Bank module...")
    
    if not MEMORY_BANK_DIR.exists():
        print_status(f"Memory-Bank module not found at {MEMORY_BANK_DIR}", False)
        return False
    
    required_files = [
        MEMORY_BANK_DIR / 'index.js',
        MEMORY_BANK_DIR / 'adapters' / 'filesystem.js'
    ]
    
    for file in required_files:
        if not file.exists():
            print_status(f"Required file {file} not found", False)
            return False
    
    print_status("Memory-Bank module verification completed")
    return True

def create_memory_storage():
    """Create memory storage directory"""
    print("Creating memory storage directory...")
    
    if ensure_directory(MEMORY_STORAGE_DIR):
        print_status(f"Memory storage directory created at {MEMORY_STORAGE_DIR}")
        return True
    return False

def create_env_file():
    """Create or update .env file with Memory-Bank configuration"""
    print("Creating/updating .env file...")
    
    env_path = PROJECT_ROOT / '.env'
    env_content = f"""# Memory-Bank Configuration
MEMORY_STORAGE_DIR="{MEMORY_STORAGE_DIR}"
VECTOR_DB_ENABLED=false
MAX_MEMORY_AGE_DAYS=30
MEMORY_COMPRESSION_ENABLED=true

# OpenRouter Configuration
# OPENROUTER_API_KEY=your_openrouter_api_key
# EMBEDDING_MODEL=openai/text-embedding-ada-002

# Optional: Supabase Vector Database
# SUPABASE_URL=your_supabase_url
# SUPABASE_KEY=your_supabase_key
# VECTOR_TABLE_NAME=memory_vectors
"""
    
    if env_path.exists():
        # Check if Memory-Bank config already exists
        with open(env_path, 'r') as f:
            existing_content = f.read()
        
        if 'MEMORY_STORAGE_DIR' not in existing_content:
            # Append Memory-Bank config to existing file
            with open(env_path, 'a') as f:
                f.write("\n# Memory-Bank Configuration\n")
                f.write(env_content)
            print_status("Updated .env file with Memory-Bank configuration")
        else:
            print_status(".env file already contains Memory-Bank configuration")
    else:
        # Create new .env file
        success = write_file(env_path, env_content)
        if success:
            print_status("Created .env file with Memory-Bank configuration")
        else:
            return False
    
    return True

def create_integration_modules():
    """Create integration modules for connecting Memory-Bank with other components"""
    print("Creating integration modules...")
    
    integration_dir = MODULES_DIR / 'integrations'
    ensure_directory(integration_dir)
    
    # Create memory-integration.js helper
    helper_path = integration_dir / 'memory-integration.js'
    helper_content = """/**
 * Memory-Bank Integration Helper
 * 
 * Provides utility functions for integrating Memory-Bank with other modules
 */

const MemoryBank = require('../memory-bank');

// Initialize memory bank
const memoryBank = new MemoryBank();

/**
 * Store data in memory bank
 * @param {string} category - Memory category
 * @param {string} key - Memory identifier
 * @param {any} data - Data to store
 * @param {object} metadata - Additional metadata
 * @returns {Promise<boolean>} Success status
 */
async function storeMemory(category, key, data, metadata = {}) {
  return await memoryBank.store(category, key, data, { metadata });
}

/**
 * Retrieve data from memory bank
 * @param {string} category - Memory category
 * @param {string} key - Memory identifier
 * @returns {Promise<any>} Retrieved data or null if not found
 */
async function retrieveMemory(category, key) {
  return await memoryBank.retrieve(category, key);
}

/**
 * Find similar memories using vector search
 * @param {string} query - Search query
 * @param {object} options - Search options
 * @returns {Promise<Array>} Array of similar memories
 */
async function findSimilarMemories(query, options = {}) {
  return await memoryBank.findSimilar(query, options);
}

/**
 * Get all memory keys in a category
 * @param {string} category - Memory category
 * @returns {Promise<string[]>} Array of memory keys
 */
async function listMemories(category) {
  return await memoryBank.list(category);
}

/**
 * Delete memory item
 * @param {string} category - Memory category
 * @param {string} key - Memory identifier
 * @returns {Promise<boolean>} Success status
 */
async function deleteMemory(category, key) {
  return await memoryBank.delete(category, key);
}

module.exports = {
  memoryBank,
  storeMemory,
  retrieveMemory,
  findSimilarMemories,
  listMemories,
  deleteMemory
};
"""
    
    if write_file(helper_path, helper_content):
        print_status(f"Created memory integration helper at {helper_path}")
    
    # Create individual integration hooks
    integration_points = [
        {
            'module': 'openrouter',
            'category': 'model-calls',
            'description': 'Manages model interaction history and cached responses',
        },
        {
            'module': 'task-master',
            'category': 'tasks',
            'description': 'Stores task progress and decision rationales',
        },
        {
            'module': 'ideas-generator',
            'category': 'ideas',
            'description': 'Preserves generated ideas and their metadata',
        }
    ]
    
    for point in integration_points:
        hook_path = integration_dir / f"{point['module']}-memory-hook.js"
        hook_content = f"""/**
 * Memory-Bank Integration Hook for {point['module']}
 * 
 * {point['description']}
 */

const {{ storeMemory, retrieveMemory, findSimilarMemories }} = require('./memory-integration');

/**
 * Store data in memory
 * @param {{}} data - Data to store
 * @param {{}} options - Storage options
 * @returns {{Promise<boolean>}} Success status
 */
async function store(data, options = {{}}) {{
  const key = options.key || generateKey(data);
  return await storeMemory('{point['category']}', key, data, options.metadata || {{}});
}}

/**
 * Retrieve data from memory
 * @param {{string}} key - Memory identifier
 * @returns {{Promise<any>}} Retrieved data
 */
async function retrieve(key) {{
  return await retrieveMemory('{point['category']}', key);
}}

/**
 * Check if similar data exists in memory
 * @param {{any}} data - Data to check
 * @param {{}} options - Search options
 * @returns {{Promise<{{exists: boolean, data: any, similarity: number}}>}} Result with similarity score
 */
async function checkSimilar(data, options = {{}}) {{
  // Convert data to query string if needed
  const query = typeof data === 'string' ? data : JSON.stringify(data);
  
  // Search for similar memories
  const results = await findSimilarMemories(query, {{
    category: '{point['category']}',
    threshold: options.threshold || 0.7,
    limit: options.limit || 5,
    ...options
  }});
  
  if (results && results.length > 0) {{
    return {{
      exists: true,
      data: results[0].data,
      similarity: results[0].score,
      results
    }};
  }}
  
  return {{ exists: false, data: null, similarity: 0, results: [] }};
}}

/**
 * Generate a deterministic key from data
 * @param {{any}} data - Input data
 * @returns {{string}} Generated key
 */
function generateKey(data) {{
  try {{
    // For objects, create a deterministic string
    if (typeof data === 'object' && data !== null) {{
      const normalized = JSON.stringify(sortObjectKeys(data));
      
      // Create a simple hash
      let hash = 0;
      for (let i = 0; i < normalized.length; i++) {{
        const char = normalized.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash; // Convert to 32bit integer
      }}
      
      return `{point['module']}-${{Math.abs(hash)}}`;
    }}
    
    // For strings, use directly with a prefix
    if (typeof data === 'string') {{
      return `{point['module']}-${{data.substring(0, 40).replace(/[^a-zA-Z0-9]/g, '-')}}`;
    }}
    
    // For other types, convert to string
    return `{point['module']}-${{String(data)}}`;
  }} catch (error) {{
    // Fallback to timestamp
    return `{point['module']}-${{Date.now()}}`;
  }}
}}

/**
 * Recursively sort object keys for deterministic JSON serialization
 * @param {{object}} obj - Object to sort
 * @returns {{object}} Sorted object
 */
function sortObjectKeys(obj) {{
  if (typeof obj !== 'object' || obj === null) {{
    return obj;
  }}
  
  if (Array.isArray(obj)) {{
    return obj.map(sortObjectKeys);
  }}
  
  return Object.keys(obj).sort().reduce((sorted, key) => {{
    sorted[key] = sortObjectKeys(obj[key]);
    return sorted;
  }}, {{}});
}}

module.exports = {{
  store,
  retrieve,
  checkSimilar,
  generateKey
}};
"""
        
        if write_file(hook_path, hook_content):
            print_status(f"Created {point['module']} integration hook at {hook_path}")
    
    return True

def update_package_json():
    """Update or create package.json with Memory-Bank dependency"""
    print("Updating package.json...")
    
    package_path = PROJECT_ROOT / 'package.json'
    package_data = {
        "name": "e2e-auto-micro-apps",
        "version": "0.1.0",
        "description": "Linux-Based AI Project Auto-Initiation System",
        "main": "index.js",
        "scripts": {
            "start": "node index.js",
            "test": "echo \"Error: no test specified\" && exit 1"
        },
        "dependencies": {},
        "devDependencies": {}
    }
    
    # Load existing package.json if it exists
    if package_path.exists():
        try:
            with open(package_path) as f:
                package_data = json.load(f)
        except json.JSONDecodeError:
            print_status(f"Invalid package.json, will create a new one", False)
    
    # Add Memory-Bank as local dependency
    if 'dependencies' not in package_data:
        package_data['dependencies'] = {}
    
    package_data['dependencies']['memory-bank'] = 'file:./modules/memory-bank'
    
    # Write updated package.json
    if write_file(package_path, json.dumps(package_data, indent=2)):
        print_status("Updated package.json with Memory-Bank dependency")
        return True
    return False

def create_main_index():
    """Create main index.js file if it doesn't exist"""
    print("Creating main index.js...")
    
    index_path = PROJECT_ROOT / 'index.js'
    
    if index_path.exists():
        print_status("Main index.js already exists, skipping")
        return True
    
    index_content = """/**
 * E2E-AUTO-MICRO-APPS
 * Linux-Based AI Project Auto-Initiation System
 * 
 * Main entry point for the orchestration pipeline
 */

// Import core modules
const MemoryBank = require('./modules/memory-bank');

// Initialize memory bank
const memoryBank = new MemoryBank();

// Export API
module.exports = {
  memoryBank,
  // Additional modules will be exported here
};

// CLI command handler (future implementation)
async function handleCommand(command, args) {
  console.log(`Command: ${command}`);
  console.log(`Arguments:`, args);
  
  switch (command) {
    case 'generate':
      console.log('Generating ideas...');
      // Call ideas generator here
      break;
      
    case 'accept':
      console.log(`Accepting idea: ${args[0]}`);
      // Call project bootstrapper here
      break;
      
    case 'list':
      console.log('Listing ideas...');
      // List ideas here
      break;
      
    case 'status':
      console.log('System status:');
      // Show system status here
      break;
      
    default:
      console.log(`Unknown command: ${command}`);
      console.log('Available commands: generate, accept, list, status');
      break;
  }
}

// Direct execution handling
if (require.main === module) {
  console.log('E2E-AUTO-MICRO-APPS orchestration system initializing...');
  
  // Process command line arguments
  const args = process.argv.slice(2);
  const command = args.shift();
  
  if (command) {
    handleCommand(command, args)
      .then(() => console.log('Command completed'))
      .catch(err => {
        console.error('Command failed:', err);
        process.exit(1);
      });
  } else {
    console.log('No command specified. Use one of: generate, accept, list, status');
  }
}
"""
    
    if write_file(index_path, index_content):
        print_status("Created main index.js file")
        return True
    return False

def main():
    """Main integration protocol"""
    print("\n=== Memory-Bank Integration Protocol ===\n")
    
    # Verify Memory-Bank module
    if not verify_memory_bank():
        print_status("Memory-Bank module verification failed, aborting integration", False)
        return 1
    
    # Create memory storage directory
    if not create_memory_storage():
        print_status("Failed to create memory storage directory, aborting integration", False)
        return 1
    
    # Create/update .env file
    if not create_env_file():
        print_status("Failed to set up environment configuration, aborting integration", False)
        return 1
    
    # Create integration modules
    if not create_integration_modules():
        print_status("Failed to create integration modules, integration may be incomplete", False)
    
    # Update package.json
    if not update_package_json():
        print_status("Failed to update package.json, integration may be incomplete", False)
    
    # Create main index.js
    if not create_main_index():
        print_status("Failed to create main index.js, integration may be incomplete", False)
    
    print("\n=== Memory-Bank Integration Complete ===\n")
    print("Integration completed successfully. The Memory-Bank module is now integrated")
    print("with the E2E-AUTO-MICRO-APPS orchestration framework.")
    print("\nUsage in JavaScript code:")
    print("const MemoryBank = require('./modules/memory-bank');")
    print("const memoryBank = new MemoryBank();")
    print("\nAPI Methods:")
    print("- memoryBank.store(category, key, data, options)")
    print("- memoryBank.retrieve(category, key)")
    print("- memoryBank.list(category)")
    print("- memoryBank.delete(category, key)")
    print("- memoryBank.findSimilar(query, options) [future implementation]")
    
    return 0

if __name__ == "__main__":
    sys.exit(main())
